src/sentry/integrations/source_code_management/repository.py

```diff
@@ -0,0 +1,200 @@
+            sentry_sdk.capture_exception()
+            return None
+        try:
+            response = client.check_file(repo, filepath, branch)
+            if response is None:
+                return None
+        except IdentityNotValid:
+            return None
+        except ApiError as e:
+            if e.code != 404:
+                sentry_sdk.capture_exception()
+                raise
+
+            return None
+
+        return self.format_source_url(repo, filepath, branch)
+
+    def get_stacktrace_link(
+        self, repo: Repository, filepath: str, default: str, version: str | None
+    ) -> str | None:
+        """
+        Handle formatting and returning back the stack trace link if the client
+        request was successful.
+
+        Uses the version first, and re-tries with the default branch if we 404
+        trying to use the version (commit sha).
+
+        If no file was found return `None`, and re-raise for non-"Not Found"
+        errors, like 403 "Account Suspended".
+        """
+        scope = sentry_sdk.Scope.get_isolation_scope()
+        scope.set_tag("stacktrace_link.tried_version", False)
+        if version:
+            scope.set_tag("stacktrace_link.tried_version", True)
+            source_url = self.check_file(repo, filepath, version)
+            if source_url:
+                scope.set_tag("stacktrace_link.used_version", True)
+                return source_url
+        scope.set_tag("stacktrace_link.used_version", False)
+        source_url = self.check_file(repo, filepath, default)
+
+        return source_url
+
+    @abstractmethod
+    def get_repositories(self, query: str | None = None) -> Sequence[dict[str, Any]]:
+        """
+        Get a list of available repositories for an installation
+
+        >>> def get_repositories(self):
+        >>>     return self.get_client().get_repositories()
+
+        return [{
+            'name': display_name,
+            'identifier': external_repo_id,
+        }]
+
+        The shape of the `identifier` should match the data
+        returned by the integration's
+        IntegrationRepositoryProvider.repository_external_slug()
+        """
+        raise NotImplementedError
+
+    def get_unmigratable_repositories(self) -> Collection[RpcRepository]:
+        """
+        Get all repositories which are in our database but no longer exist as far as
+        the external service is concerned.
+        """
+        return []
+
+    def reinstall_repositories(self) -> None:
+        """Reinstalls repositories associated with the integration."""
+        result = integration_service.organization_contexts(integration_id=self.model.id)
+
+        for install in result.organization_integrations:
+            repository_service.reinstall_repositories_for_integration(
+                organization_id=install.organization_id,
+                integration_id=self.model.id,
+                provider=f"integrations:{self.model.provider}",
+            )
```

Comment:
Should this be wrapped in a try-catch, or are we okay with only some of the repositories being resintalled?

Reply:
i copied this from the old mixin, this is wrapped in an atomic transaction?

Reply:
nvm, seems like this is unused. removed!